paymentsheet/src/main/java/com/stripe/android/paymentsheet/flowcontroller/DefaultFlowController.kt

```diff
@@ -264,45 +266,77 @@ internal class DefaultFlowController @Inject internal constructor(
         paymentSelection: PaymentSelection?,
         state: PaymentSheetState.Full,
     ) {
-        val mode = viewModel.initializationMode ?: return
+        lifecycleScope.launch {
+            val stripeIntent = requireNotNull(state.stripeIntent)
 
-        val clientSecret = when (mode) {
-            is PaymentSheet.InitializationMode.PaymentIntent -> {
-                PaymentIntentClientSecret(mode.clientSecret)
-            }
-            is PaymentSheet.InitializationMode.SetupIntent -> {
-                SetupIntentClientSecret(mode.clientSecret)
-            }
-            is PaymentSheet.InitializationMode.DeferredIntent -> {
-                TODO("Not implemented yet")
+            val nextStep = intentConfirmationInterceptor.intercept(
+                clientSecret = stripeIntent.clientSecret,
+                paymentSelection = paymentSelection,
+                shippingValues = state.config?.shippingDetails?.toConfirmPaymentIntentShipping(),
+            )
+
+            when (nextStep) {
+                is IntentConfirmationInterceptor.NextStep.HandleNextAction -> {
+                    handleNextAction(
+                        clientSecret = nextStep.clientSecret,
+                        stripeIntent = stripeIntent,
+                    )
+                }
+                is IntentConfirmationInterceptor.NextStep.Confirm -> {
+                    confirmStripeIntent(nextStep.confirmParams)
+                }
+                is IntentConfirmationInterceptor.NextStep.Fail -> {
+                    onPaymentResult(
+                        PaymentResult.Failed(
+                            // TODO (jameswoo): Exposing new type here
+                            ConfirmPaymentException(nextStep.errorMessage)
+                        )
+                    )
```

Comment:
Wouldn't they know from the type? How else would we get more technical data on why the confirm failed if it was passed from the merchant?

Reply:
It could fail for a variety of reasons (creating the payment method failed, retrieving the intent failed, or something else). I’m wondering if we’d want to relay that to the merchant so that they can log it for their debug builds.

Reply:
Should we expose some new exception types, or should we just expose a new developer error message?

Reply:
I think we could add an `exception` field to `NextStep.Fail`.

- For failures in `retrieveStripeIntent` and `createPaymentMethod`, we can just use the caught exception.
- For the case where `createIntentCallback` returns a failure, we could put convert it to a `ConfirmPaymentException` (which should probably be called `CreateIntent(Callback)Exception`) and return that.

In the custom flow, the merchant can log the `CreateIntentException`, but should use a user-facing message of their own (as they are already force to do today).